### PR TITLE
Implementation of Gumbel Distribution

### PR DESCRIPTION
**Pull request by vishwakftw**
Math checked and approved by @fritzo .

cc: @apaszke

### CODE REVIEW

File: torch/distributions/gumbel.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        else:
+            batch_shape = self.scale.size()
+        super(Gumbel, self).__init__(batch_shape)
+
+    def rsample(self, sample_shape=torch.Size()):
+        shape = self._extended_shape(sample_shape)
+        uni_dist = self.scale.new(shape).uniform_()
+        # X ~ Uniform(0, 1)
+        # Y = loc - scale * ln (-ln (X)) ~ Gumbel(loc, scale)
+        return self.loc - self.scale * torch.log(-uni_dist.log())
```
Comment (vadimkantorov) [inline torch/distributions/gumbel.py]: Can this produce `-inf` if `unform_` happens to sample exactly 0?

Comment (vishwakftw) [reply torch/distributions/gumbel.py]: Yes, you are right. In `aten/src/TH/THRandom.c` the uniform double is generated in the interval [0, 1). I will try to fix this ASAP.

Comment (vishwakftw) [reply torch/distributions/gumbel.py]: I have fixed it and the quantiles are looking good and consistent with the ones in `scipy.stats.gumbel_r`. Thank you for raising this issue.

**Comment by yf225** [comment]
Windows CI seems to be stuck at cuda test which should be unrelated. Retesting.

File: torch/distributions/gumbel.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 from numbers import Number
 
 import math
+from numpy import euler_gamma
```
Comment (vadimkantorov) [inline torch/distributions/gumbel.py]: This introduces a hard depenendcy on NumPy. Is it desired?

Comment (fritzo) [reply torch/distributions/gumbel.py]: No we should simply hard code this number:
```py
euler_gamma =  0.57721566490153286060
```

Comment (vishwakftw) [reply torch/distributions/gumbel.py]: Ah, yes. You are right about the hard dependency. But, I suppose NumPy is required for PyTorch. @apaszke could you suggest what should be done?

File: torch/distributions/gumbel.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        # Y = loc - scale * ln (-ln (X)) ~ Gumbel(loc, scale)
+        return self.loc - self.scale * torch.log(-uni_dist.log())
+
+    def log_prob(self, value):
+        self._validate_log_prob_arg(value)
+        z = (value - self.loc) / self.scale
+        return -(self.scale.log() + z + torch.exp(-z))
+
+    def entropy(self):
+        return self.scale.log() + 1 + euler_gamma
```
Comment (fritzo) [inline torch/distributions/gumbel.py]: nit: parens around the floats would save a tensor op: `(1 + euler_gamma)`

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Thanks @vishwakftw!

**Pull request merged**